synapse/storage/stream.py

```diff
@@ -127,6 +129,91 @@ def upper_bound(self):
 
 
 class StreamStore(SQLBaseStore):
+
+    @defer.inlineCallbacks
+    def get_appservice_room_stream(self, service, from_key, to_key, limit=0):
+        # NB this lives here instead of appservice.py so we can reuse the
+        # 'private' StreamToken class in this file.
+        if limit:
+            limit = max(limit, MAX_STREAM_SIZE)
+        else:
+            limit = MAX_STREAM_SIZE
+
+        # From and to keys should be integers from ordering.
+        from_id = _StreamToken.parse_stream_token(from_key)
+        to_id = _StreamToken.parse_stream_token(to_key)
+
+        if from_key == to_key:
+            defer.returnValue(([], to_key))
+            return
+
+        # Logic:
+        #  - We want ALL events which match the AS room_id regex
+        #  - We want ALL events which match the rooms represented by the AS
+        #    room_alias regex
+        #  - We want ALL events for rooms that AS users have joined.
+        # This is currently supported via get_app_service_rooms (which is used
+        # for the Notifier listener rooms). We can't reasonably make a SQL
+        # query for these room IDs, so we'll pull all the events between from/to
+        # and filter in python.
+        rooms_for_as = yield self.get_app_service_rooms(service)
+        room_ids_for_as = [r.room_id for r in rooms_for_as]
+
+        # select all the events between from/to with a sensible limit
+        sql = (
+            "SELECT e.event_id, e.room_id, e.type, e.unrecognized_keys, "
+            "e.stream_ordering FROM events AS e "
+            "WHERE e.stream_ordering > ? AND e.stream_ordering <= ? "
+            "ORDER BY stream_ordering ASC LIMIT %(limit)d "
+        ) % {
+            "limit": limit
+        }
+
+        def app_service_interested(row):
+            if row["room_id"] in room_ids_for_as:
+                return True
+
+            if row["type"] == EventTypes.Member:
+                # load up the content to inspect if some user the AS is
+                # interested in was invited to a room. We'll be passing this
+                # through _get_events_txn later, so ignore the fact that this
+                # may be a redacted event.
+                event_content = simplejson.loads(row["unrecognized_keys"])
+                if (service.is_interested_in_user(
+                        event_content.get("state_key"))):
+                    return True
```

Comment:
Probably `state_events`, but I'm not sure I have too much of an opinion on that.